File: 백준/Silver/#1260 DFS와 BFS/DFS와 BFS.py
import sys

input = sys.stdin.readline



def dfs(current_node):
    global graph, dfs_visited, dfs_path

    dfs_visited[current_node] = 1
    dfs_path.append(current_node)

    for next_node in graph[current_node]:
        if not dfs_visited[next_node]:
            dfs(next_node)

def bfs(start_node):
    global graph, bfs_visited, bfs_path

    queue = list()
    queue.append(start_node)
    bfs_path.append(start_node)

    bfs_visited[start_node] = 1

    while len(queue):
        current_node = queue[0]
        del queue[0]
        
        for next_node in graph[current_node]:
            if not bfs_visited[next_node]:
                queue.append(next_node)
                bfs_path.append(next_node)
                bfs_visited[next_node] = 1




n, m, start_node = map(int, input().split())

graph = [[] for i in range(n + 1)]
dfs_visited = [ 0 for i in range(n + 1)]
bfs_visited = [ 0 for i in range(n + 1)]

dfs_path = list()
bfs_path = list()

for i in range(m):
    u, v = map(int, input().split())
    graph[u].append(v)
    graph[v].append(u)

for i in range(n + 1):
    graph[i].sort()

dfs(start_node)
bfs(start_node)
print(*dfs_path)
print(*bfs_path)

File: 백준/Silver/#1260 DFS와 BFS/DFS와 BFS.cc
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>
#include <climits>
#include <queue>

using namespace std;

const int MAX = 10001;
int visited[MAX];
int n, m, u, v, startNode;

int dx[] = {0, 0, 1, -1};
int dy[] = {1, -1, 0, 0};

void dfs(vector<vector<int>> &adjList, int currentNode) {
	visited[currentNode] = 1;
	cout << currentNode << " ";

	for (int nextNode: adjList[currentNode]) {
		if (visited[nextNode]) continue;
		dfs(adjList, nextNode); 
	}
	return;
}

void bfs(vector<vector<int>> &adjList) {

	queue<int> q;
	q.push(startNode);
	visited[startNode] = 1;
	cout << startNode << " ";

	while(!q.empty()) {
		int currentNode = q.front();
		q.pop();

		for (int nextNode: adjList[currentNode]) {
			if (visited[nextNode] == 0) {
				q.push(nextNode);
				visited[nextNode] = 1;
				cout << nextNode << " ";
			}
		}
	}
}


int main() {

	ios_base::sync_with_stdio(false);
    cout.tie(nullptr);
    cin.tie(nullptr);
	
	cin >> n >> m >> startNode;

	vector<vector<int>> adjList (n + 1, vector<int>());

	for (int i = 0; i < m; i++) {
		cin >> u >> v;
		adjList[u].push_back(v);
		adjList[v].push_back(u);
	}

	for (int i = 1; i <= n; i++) {
		sort(adjList[i].begin(), adjList[i].end());
	}

	fill(visited, visited + n + 1, 0);
	dfs(adjList, startNode);
	cout << endl;
	fill(visited, visited + n + 1, 0);
	bfs(adjList);

    return 0;
}
